rules/active_ruleset.go

```diff
@@ -0,0 +1,745 @@
+					CutoverNanos: snapshot.cutoverNanos,
+					Tombstoned:   false,
+					Metadata:     metadata.Metadata{Pipelines: filteredPipelines},
+				}, true
+			}
+		}
+	}
+	return metadata.DefaultStagedMetadata, false
+}
+
+// nextCutoverIdx returns the next snapshot index whose cutover time is after t.
+// NB(xichen): not using sort.Search to avoid a lambda capture.
+func (as *activeRuleSet) nextCutoverIdx(t int64) int {
+	i, j := 0, len(as.cutoverTimesAsc)
+	for i < j {
+		h := i + (j-i)/2
+		if as.cutoverTimesAsc[h] <= t {
+			i = h + 1
+		} else {
+			j = h
+		}
+	}
+	return i
+}
+
+// cutoverNanosAt returns the cutover time at given index.
+func (as *activeRuleSet) cutoverNanosAt(idx int) int64 {
+	if idx < len(as.cutoverTimesAsc) {
+		return as.cutoverTimesAsc[idx]
+	}
+	return timeNanosMax
+}
+
+// filterByAggregationType takes a list of pipelines as input and returns those
+// containing the given aggregation type.
+func filteredPipelinesWithAggregationType(
+	pipelines []metadata.PipelineMetadata,
+	mt metric.Type,
+	at aggregation.Type,
+	opts aggregation.TypesOptions,
+) []metadata.PipelineMetadata {
+	var cur int
+	for i := 0; i < len(pipelines); i++ {
+		var containsAggType bool
+		if aggID := pipelines[i].AggregationID; aggID.IsDefault() {
+			containsAggType = opts.IsContainedInDefaultAggregationTypes(at, mt)
+		} else {
+			containsAggType = aggID.Contains(at)
+		}
+		if !containsAggType {
+			continue
+		}
+		if cur != i {
+			pipelines[cur] = pipelines[i]
+		}
+		cur++
+	}
+	return pipelines[:cur]
+}
+
+// mergeResultsForNewRollupIDs merges the current list of staged metadatas for new rollup IDs
+// with the list of staged metadatas for new rollup IDs at the next rule cutover time, assuming
+// that both the current metadatas list and the next metadatas list are sorted by rollup IDs
+// in ascending order.
+// NB: each item in the `nextResults` array has a single staged metadata in the `metadatas` array
+// as the staged metadata for the associated rollup ID at the next cutover time.
+func mergeResultsForNewRollupIDs(
+	currResults []IDWithMetadatas,
+	nextResults []IDWithMetadatas,
+	nextCutoverNanos int64,
+) []IDWithMetadatas {
+	var (
+		currLen, nextLen = len(currResults), len(nextResults)
+		currIdx, nextIdx int
+	)
+	for currIdx < currLen || nextIdx < nextLen {
+		var compareResult int
+		if currIdx >= currLen {
+			compareResult = 1
+		} else if nextIdx >= nextLen {
+			compareResult = -1
+		} else {
+			compareResult = bytes.Compare(currResults[currIdx].ID, nextResults[nextIdx].ID)
+		}
+
+		// If the current result and the next result have the same ID, we append the next metadata
+		// to the end of the metadata list.
+		if compareResult == 0 {
+			currResults[currIdx].Metadatas = append(currResults[currIdx].Metadatas, nextResults[nextIdx].Metadatas[0])
+			currIdx++
+			nextIdx++
+			continue
+		}
+
+		// If the current ID is smaller, it means the current rollup ID is tombstoned at the next
```

Comment:
So the matching logic goes like this:

Imagine you have the list of rules in the ruleset lined up vertically. Each rule may have one or more snapshots, each of which represents a change to that rule (e.g., filter change, policy change, etc.). These snapshots are naturally non-overlapping in time since only one snapshot can be active at a given point in time. As a result, if we use the x axis to represent time, then for each rule, a snapshot is active for some amount of time. IOW, if you pick a time and draw a vertical line across the set of rules, the snapshots of different ruels that intersect with the vertical line are the active rule snapshots for the ruleset.

Now you have a list of times you need to perform rule matching at. Each matching time corresponds to a cutover time of a rule in the ruleset, because that's when matching the metric ID against this rule may lead to a different metadata including different storage policies and new rollup IDs to be generated or existing rollup IDs to stop being generated. The final match result is a collection of such metadata sorted by time in ascending order.

Now going back to the comment, basically the current ID refers to one of the rollup IDs that are present at the "current" time but no longer present at the "next" time, which means the rollup ID should stop being generated at the next cutover time and as such should be marked as tombstoned.

Hopefully that makes it clear.

Reply:
Yeah, make sense, thanks for the nice explanation, I think this should being added as a README in the repo to explain the rule matching logic.

Reply:
I'll add them as comments, README might be a bit overkill for this purpose